core/client/fs/src/main/java/alluxio/client/file/ufs/UfsFileInStream.java

```diff
@@ -127,15 +134,38 @@ public void seek(long pos) throws IOException {
     if (mPosition == pos) {
       return;
     }
-    if (mPosition < pos) {
-      mPosition += mUfsInStream.skip(pos - mPosition);
+    if (!mUfsInStream.isPresent()) {
+      mPosition = pos;
+      return;
+    }
+    if (mUfsInStream.get() instanceof SeekableUnderFileInputStream) {
+      ((SeekableUnderFileInputStream) mUfsInStream.get()).seek(pos);
+    } else if (mPosition < pos) {
+      while (mPosition < pos) {
+        mPosition += mUfsInStream.get().skip(pos - mPosition);
+      }
     } else {
-      mUfsInStream.reset();
-      mPosition = mUfsInStream.skip(pos);
+      closeStream();
```

Comment:
any reason we change behavior here although we should not reach here?

Reply:
We want to support seek() even when the underlying UFS stream does not support seek backward. So we close the stream here, following read operation will reopen the stream with the new updated seek position.